pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,172 @@
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		socketTwDesc: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		socketMemoryDesc: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+type socketStats struct {
+	tcpInuse uint64 // num of tcp inuse sockets
+	udpInuse uint64 // num of udp inuse sockets
+	tcpTw    uint64 // num of tcp tw sockets
+	mem      uint64 // sum of tcp and udp memory
+}
+
+func getSockStats(fileName string) (*socketStats, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func parseSockStats(r io.Reader) (*socketStats, error) {
+	var ss socketStats
+	var lineCount int = 0
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		lineCount++
+		in := scanner.Text()
+		line := strings.Split(in, " ")
+		if len(line) < 2 || len(line[0]) < 2 || line[0][len(line[0])-1] != ':' {
+			return nil, fmt.Errorf("invalid input %q", in)
+		}
+		// Remove trailing ':'.
+		protocol := line[0][:len(line[0])-1]
+		if protocol == "TCP" || protocol == "TCP6" {
+			for i := 1; i < len(line) && i+1 < len(line); i += 2 {
+				switch line[i] {
+				case "inuse":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.tcpInuse = v
+					}
+				case "tw":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.tcpTw = v
+					}
+				case "mem":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						// memeory is summed across TCP and UDP
+						ss.mem += v
+					}
+				}
+			}
+
+		} else if protocol == "UDP" || protocol == "UDP6" {
+			for i := 1; i < len(line) && i+1 < len(line); i += 2 {
+				switch line[i] {
+				case "inuse":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						ss.udpInuse = v
+					}
+				case "mem":
+					v, err := strconv.ParseUint(line[i+1], 10, 64)
+					if err != nil {
+						return nil, fmt.Errorf("invalid value %q in sockstats: %v", line[i+1], err)
+					} else {
+						// memeory is summed across TCP and UDP
+						ss.mem += v
+					}
+				}
+			}
+
+		}
+	}
+	if lineCount == 0 {
+		return nil, fmt.Errorf("Empty file")
+	}
+	return &ss, nil
+}
+
+func (c *sockStatCollector) Update(ch chan<- prometheus.Metric) error {
+
+	ss, err := getSockStats(procFilePath("net/sockstat"))
+	if err != nil {
+		return fmt.Errorf("couldn't get sockstats: %v", err)
```

Comment:
preference is to not wrap errors unless there is a good reason
`return err`

Reply:
Updated